### PR TITLE
Consolidate nth / last object Groupby Implementations

### LINKED ISSUE #19569: CLN: consolidate last / nth object dtype groupby function implementation
ideally consolidate these with the templated implementation in ``groupby_helper.pxi.in``

### PR DESCRIPTION
**Pull request by WillAyd**
- [X] closes #19569
- [X] tests added / passed
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

Depending on what goes first between this and #19481 some merge conflicts will need to be cleaned up

### CODE REVIEW

File: pandas/_libs/groupby_helper.pxi.in [hunk 1]
```diff
@@ -360,11 +361,19 @@ def group_last_{{name}}(ndarray[{{dest_type2}}, ndim=2] out,
         raise AssertionError("len(index) != len(labels)")
 
     nobs = np.zeros((<object> out).shape, dtype=np.int64)
+    {{if name=='object'}}
+    resx = np.empty((<object> out).shape, dtype=object)
```
Comment (WillAyd) [inline pandas/_libs/groupby_helper.pxi.in]: Was really hoping to not even have this conditional, but when trying `resx = np.empty_like(out)` and even `resx = np.empty_like(out, dtype='object')` it kept SegFaulting on objects

File: pandas/tests/groupby/test_groupby.py [hunk 2]
```diff
@@ -2086,47 +2086,64 @@ def test_median_empty_bins(self):
...
 
-                grpd = df_in.groupby('a')
-                t = getattr(grpd, method)(*data['args'])
-                assert_frame_equal(t, df_out)
+    @pytest.mark.parametrize("method,exp,args", [
+        ('first', [('bar', 'quuz'), ('foo', 'baz')], []),
+        ('last', [('bar', 'grault'), ('foo', 'quux')], []),
+        ('nth', [('bar', 'corge'), ('foo', 'qux')], [1]),
+    ])
+    def test_groupby_get_nth_object(self, method, exp, args):
```
Comment (jreback) [inline pandas/tests/groupby/test_groupby.py]: i think the nth tests are in another test file

File: pandas/tests/groupby/test_groupby.py [hunk 3]
```diff
@@ -2086,47 +2086,45 @@ def test_median_empty_bins(self):
         expected = df.groupby(bins).agg(lambda x: x.median())
         assert_frame_equal(result, expected)
```
Comment (jreback) [inline pandas/tests/groupby/test_groupby.py]: can you add a tests with the numeric ops on groupby / object and assert they raise

**Review by jreback — CHANGES_REQUESTED** [review]
lgtm. going to merge the other PR first.

File: pandas/tests/groupby/test_groupby.py (hunk 3 above)
Comment (WillAyd) [reply pandas/tests/groupby/test_groupby.py]: Assuming you want this in a different change no problem. That said, do you consider the numeric ops to be `add`, `prod`, `min`, `max`, `mean`, `median`, `var`, `ohlc`, `cumprod`, `cumsum`, `cummin`, `cummax` and `rank`? The tests are one thing, but these don't all raise at the moment so would have to couple that with some refactoring of the `groupby` module

Comment (jreback) [reply pandas/tests/groupby/test_groupby.py]: yes let's do this in another change.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks!

**Comment by jreback** [comment]
@WillAyd I merged this, then reverted it. Can you send a new PR that is rebased on master?

**Comment by jreback** [comment]
Reverted in https://github.com/pandas-dev/pandas/commit/5c76f33a106d071e3b5620f908cfc760d8e3d6b2

**Pull request merged**